### ISSUE TITLE
Poor CPU performance.

### ISSUE THREAD

**Issue by zet4** [go-flutter-desktop/go-flutter#59]
###### `$ flutter --version`

```
Flutter 1.0.0 • channel stable • https://github.com/flutter/flutter.git
Framework • revision 5391447fae (6 weeks ago) • 2018-11-29 19:41:26 -0800
Engine • revision 7375a0f414
Tools • Dart 2.1.0 (build 2.1.0-dev.9.4 f9ebf21297)
```

###### Download [Link](https://github.com/Drakirus/go-flutter-desktop-embedder/releases)
 - Yes

## Steps to Reproduce

1. Run any flutter app (including the samples provided with the project)
2. Observe abnormal CPU usage (10-15% on idle)

## Notes

I have not tested google's version of this as I never got it to build correctly. Hence I am not sure if this is a normal behaviour for flutter on desktop, but I am hoping it's not.

I have also attempted to implement FPS limit in the "game" loop of GLFW, but that didn't do anything to throttle this thing's greed for computing.

**Comment by Drakirus** [comment]
https://github.com/google/flutter-desktop-embedding/issues/58

**Comment by zet4** [comment]
So I finally got google/fde to build and run on my windows machine and the performance / resource usage was a lot lower than what I saw with Go bindings here, so it's not completely on Flutter here, is there anything we can do to figure out what's causing the constant 15% on idle?

**Comment by Drakirus** [comment]
On which platform are you running ?

**Comment by zet4** [comment]
Both of the embedders were tested on Windows 10 x64. With `go version go1.11rc1`, which is a bit behind, but not enough to cause any drastic changes, the sample builds in gh releases were showing the same results anyway.

**Comment by Drakirus** [comment]
Using `glfw.PollEvents()` in [gutter.go](https://github.com/Drakirus/go-flutter-desktop-embedder/blob/f22c94dc7facb2b3e53571ea0e7d7b3103140c38/gutter.go#L47-L51) is way more CPU intensive than using `glfw.WaitEvents()`.
If you want the embedder to receive a platform plugin message from the flutter side. The line `flutter.EngineFlushPendingTasksNow()` has to be run ([flutter#16912](https://github.com/flutter/flutter/issues/16912)).

When using `glfw.PollEvents()` the line `flutter.EngineFlushPendingTasksNow()` is always called (infinite loop). If we use `glfw.WaitEvents()` (The best solution to improve the resource usage ) there has to be a new frame to be rendered to notify the embedder of a new platform message.

I'm using `glfw.PollEvents()` for the demo because in some edge cases the platform message was not received by the embedder, making the keyboard plugin fail.

**Comment by Drakirus** [comment]
The Windows version of **flutter-desktop-embedding** seems to be using [`glfwWaitEvents`](https://github.com/google/flutter-desktop-embedding/blob/123fea23efcf444d560df2424fa7860eab13d45b/library/common/glfw/embedder.cc#L359-L366)

**Comment by zet4** [comment]
Alright awesome, after switching to WaitEvents the CPU usage has become tolerable for my use case. 
As you mentioned using this method has its drawbacks too, keyboard plugin in particular, even though I've yet to notice said edge cases from my short testing, but since my project will likely not require much keyboard input, I will hope it's going to be okay for me.

I am assuming nothing else can really be done about this until 16912 is resolved.

**Comment by thetooth** [comment]
Just bumping this as I am experiencing similar issues. Are you saying 15% total CPU or single core utilization? Because I am getting 100% single thread utilization that is locked to one core (with the same environment using the simple demo example). 

Unfortunately trying to use this in an embedded setting(Intel NUC) won't end well :hot_pepper: :hot_pepper: :hot_pepper:

I have tried switching to WaitEvents and this completely solves the issue, I am also not seeing the edge case mentioned.

**Comment by GeertJohan** [comment]
I have included a change in #86 that should imrpove this issue by limiting the message loop to ~60fps iterations.

**Comment by GeertJohan** [comment]
#86 has been merged.The performence is not yet 100% efficient, but seems to be the best we can do with the currently available features from engine/embedder. It's certainly not "poor performance" anymore, so I'm closing this issue.
